### PR TITLE
Feat/ add CI-EX connector

### PR DESCRIPTION
**Pull request by CrimsonJacket**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Adds CI-EX connector into Foundation

[sc-35018]

**Tests performed by the developer**:
Modified `ExchangePyBase` to allow for connectors that lacks `UserStreamDataSource`.
Adds all connector components for CI-EX

**Tips for QA testing**:
Ensure that the connector runs as intended.

Note: CI-EX API does not support any private user streams, as such it is expected that order and trade updates are not received immediately.

### CODE REVIEW

**Comment by Deepali6997** [comment]
testrun : https://hummingbot.testrail.io/index.php?/runs/view/3201

**Review by RC-13 — APPROVED** [review]
LGTM

Test summary:
        * Install feature branch without issue
        * Start the client
        * Add an API key (invalid and valid key): Ok
        * Balance check: Ok
        * Create a PMM strategy and confirmed market list is shown
        * Started the strategy without any issue - close linked GitHub ticket
        * Data integrity: OK
        * Order creation and cancelation: Ok, however we have an expected behavior wherein we can experience delay as we send request using HTTP request.
        * Enable multiple order levels: Ok
        * Enable hanging orders: Ok
        * Setup multiple bot (2 bot on this test due to limited test asset): Ok
        * Create a basic TWAP strategy using CIEX connector and start: Ok
        * Build a docker image and start a docker instance
        * Run a basic PMM strategy using CIEX connector

**Review by aarmoa — CHANGES_REQUESTED** [review]

File: hummingbot/connector/exchange/ciex/ciex_exchange.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+        return self._trading_required
+
+    def _expected_initial_status_dict(self) -> Dict[str, bool]:
+        return {
+            "symbols_mapping_initialized": False,
+            "order_books_initialized": False,
+            "account_balance": False,
+            "trading_rule_initialized": False,
+            "user_stream_initialized": True,
+        }
```
[the remark this answers is not in the document]
Comment (CrimsonJacket) [reply hummingbot/connector/exchange/ciex/ciex_exchange.py:111]: Removed

File: hummingbot/connector/exchange/ciex/ciex_exchange.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        is_time_synchronizer_related = "HTTP status is 429" in error_description
+        return is_time_synchronizer_related
+
+    def _is_request_result_an_error_related_to_time_synchronizer(self, request_result: Dict[str, Any]) -> bool:
+        return request_result.get("code") == CONSTANTS.INVALID_TIMESTAMP_ERROR_CODE
+
+    def _is_order_not_found_during_status_update_error(self, status_update_exception: Exception) -> bool:
+        return str(CONSTANTS.ORDER_DOES_NOT_EXIST_ERROR_CODE) in str(
+            status_update_exception
+        ) and CONSTANTS.ORDER_NOT_EXIST_MESSAGE in str(status_update_exception)
```
[the remark this answers is not in the document]
Comment (CrimsonJacket) [reply hummingbot/connector/exchange/ciex/ciex_exchange.py:188]: Removed.

File: test/hummingbot/connector/exchange/ciex/test_ciex_exchange.py [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+                    "time": 1499865549590,
+                    "isBuyer": True,
+                    "isMaker": True,
+                    "feeCoin": self.expected_partial_fill_fee.flat_fees[0].token,
+                    "fee": float(self.expected_partial_fill_fee.flat_fees[0].amount),
+                }
+            ]
+        }
+
+    # TODO: Add tests that updates the order statuses when a subsequent order status update interval is reached
```
[the remark this answers is not in the document]
Comment (CrimsonJacket) [reply test/hummingbot/connector/exchange/ciex/test_ciex_exchange.py:674]: No, I was implementing it, but realised that it was essentially covered by existing unit tests in `ExchangeConnectorTests`

**Review by aarmoa — APPROVED** [review]

**Comment by CrimsonJacket** [comment]
Closing this PR. I've created another PR on Foundation [here](https://github.com/hummingbot/hummingbot/pull/5959)

**Pull request closed without merging**